src/main/java/com/ibm/devops/connect/CRPipeline/UploadDeployment.java

```diff
@@ -0,0 +1,272 @@
+        String startTime = envVars.expand(this.startTime);
+        String endTime = envVars.expand(this.endTime);
+
+        JSONObject payload = new JSONObject();
+
+        // user-provided inputs
+        payload.put("tenant_id", tenantId);
+        if (id != null && !id.equals("")) {
+            payload.put("id_external", id);
+        } else {
+            payload.put("id_external", versionExtId);
+        }
+        if (versionName != null && !versionName.equals("")) {
+            payload.put("version_name", versionName);
+        }
+        if (versionExtId != null && !versionExtId.equals("")) {
+            payload.put("version_id_external", versionExtId);
+        }
+        payload.put("type", type);
+        payload.put("environment_id", environmentId);
+        payload.put("environment_name", environmentName);
+        if (description != null && !description.equals("")) {
+            payload.put("description", description);
+        }
+        JSONObject application = new JSONObject();
+        if (appId != null && !appId.equals("")) {
+            application.put("id", appId);
+        }
+        if (appName != null && !appName.equals("")) {
+            application.put("name", appName);
+        }
+        if (appExtId != null && !appExtId.equals("")) {
+            application.put("externalId", appExtId);
+        }
+        if (application.isEmpty()) {
+            throw new RuntimeException("Must specify at least one of: 'appId', 'appName', 'appExtId'");
+        }
+        payload.put("application", application);
+
+        // user-provided inputs with fallbacks
+        if (name != null && !name.equals("")) {
+            payload.put("name", name);
+        } else {
+            payload.put("name", build.getParent().getName());
+        }
+        if (initiator != null && !initiator.equals("")) {
+            payload.put("by_user", initiator);
+        } else {
+            for (Cause cause : build.getCauses()) {
+                if (cause instanceof UserIdCause) {
+                    UserIdCause userCause = (UserIdCause)cause;
+                    payload.put("by_user", userCause.getUserName());
+                } else if (cause instanceof UpstreamCause) {
+                    UpstreamCause upstreamCause = (UpstreamCause)cause;
+                    payload.put("by_user", "Upstream job \"" + upstreamCause.getUpstreamProject() + "\", build \"" + upstreamCause.getUpstreamBuild() + "\"");
+                } else if (cause instanceof RemoteCause) {
+                    RemoteCause remoteCause = (RemoteCause)cause;
+                    payload.put("requestor", remoteCause.getAddr());
+                } else {
+                    payload.put("requestor", cause.getShortDescription());
+                }
+            }
+        }
+        if (succeeded != null && !succeeded.equals("")) {
+            payload.put("result", succeeded.equals("true") ? "Success" : "Failed");
```

Comment:
the build upload uses lower case statuses, which is correct? (this applies to a few lines below as well)